synapse/lib/stormlib/vault.py

```diff
@@ -0,0 +1,649 @@
+                mesg = f'User {user.name} cannot create {scope} vaults.'
+                raise s_exc.AuthDeny(mesg=mesg, user=user)
+
+            if scope == 'user' and user.iden != owner and not user.isAdmin():
+                mesg = f'User {user.name} cannot create vaults for user {owner.name}.'
+                raise s_exc.AuthDeny(mesg=mesg)
+
+        vault = {
+            'name': name,
+            'type': vtype,
+            'scope': scope,
+            'owner': owner,
+            'data': data,
+        }
+
+        return await self.runt.snap.core.addVault(vault)
+
+    async def _getByName(self, name):
+        name = await s_stormtypes.tostr(name)
+
+        vault = self.runt.snap.core.reqVaultByName(name)
+        self._reqEasyPerm(vault, s_cell.PERM_READ)
+
+        return Vault(self.runt, vault.get('iden'))
+
+    async def _getByIden(self, iden):
+        iden = await s_stormtypes.tostr(iden)
+
+        vault = self.runt.snap.core.reqVault(iden)
+        self._reqEasyPerm(vault, s_cell.PERM_READ)
+
+        return Vault(self.runt, iden)
+
+    async def _getByType(self, vtype, scope=None):
+        vtype = await s_stormtypes.tostr(vtype)
+        scope = await s_stormtypes.tostr(scope, noneok=True)
+
+        vault = self.runt.snap.core.getVaultByType(vtype, self.runt.user.iden, scope)
+        self._reqEasyPerm(vault, s_cell.PERM_READ)
+
+        return Vault(self.runt, vault.get('iden'))
+
+    async def _listVaults(self):
+        for vault in self.runt.snap.core.listVaults():
+            if not self.runt.snap.core._hasEasyPerm(vault, self.runt.user, s_cell.PERM_READ):
+                continue
+
+            yield Vault(self.runt, vault.get('iden'))
+
+@s_stormtypes.registry.registerType
+class VaultData(s_stormtypes.Prim):
+    '''
+    Implements the Storm API for Vault data.
+
+    Callers (instantiation) of this class must have already checked that the user has at least
+    PERM_EDIT to the vault.
+    '''
+    _storm_typename = 'vault:data'
+    _ismutable = False
+
+    def __init__(self, runt, valu, path=None):
+        s_stormtypes.Prim.__init__(self, valu, path=path)
+        self.runt = runt
+
+    @s_stormtypes.stormfunc(readonly=False)
+    async def setitem(self, name, valu):
+        vault = self.runt.snap.core.reqVault(self.valu)
+
+        name = await s_stormtypes.tostr(name)
+
+        if valu is s_stormtypes.undef:
+            valu = s_common.novalu
+        else:
+            valu = await s_stormtypes.toprim(valu)
+
+        return await self.runt.snap.core.setVaultData(self.valu, name, valu)
+
+    async def deref(self, name):
+        vault = self.runt.snap.core.reqVault(self.valu)
+
+        name = await s_stormtypes.tostr(name)
+
+        data = vault.get('data')
+        valu = data.get(name, s_common.novalu)
+
+        if valu is not s_common.novalu:
+            return valu
+
+        raise s_exc.NoSuchName(mesg=f'Cannot find name [{name}]', name=name, styp=self.__class__.__name__)
```

Comment:
dict style access in storm normally would return `None` on a empty / invalid key.

The docstring for this class should be updated to note that de-referencing a value which does not exist may raise a NoSuchName error.